In [18]:
import pandas as pd

# Problem Set 1

Download the dataset train.csv posted with this Exercise. This dataset is collected from Kaggle. You may also want to
download it directly from the following Kaggle Competition (Login > Go to “Data” > “Download All” > train.csv). Either
way, read the competition description (no login required) to get an idea about what the target Data Science task is.

Source : Kaggle Competition : House Prices : [https://www.kaggle.com/c/house-prices-advanced-regression-techniques](https://www.kaggle.com/c/house-prices-advanced-regression-techniques)

## Problem A

Import the “train.csv” data you downloaded (either from NTU Learn or Kaggle) in Jupyter Notebook.

In [19]:
train_csv_data = pd.read_csv("train.csv")
train_csv_data.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities  ... PoolArea PoolQC Fence MiscFeature MiscVal MoSold  \
0         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
1         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      5   
2         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      9   
3         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
4         Lvl    AllPub  ...        0    NaN   NaN         NaN       0     12   

  YrSold  SaleType  SaleCondition  SalePrice  
0   2008        WD         Normal     208500  
1   2007        WD         Normal     181500  
2   2008        WD         Normal     223500  
3   2006        WD        Abnorml     140000  
4   2008        WD         Normal     250000  

[5 rows x 81 columns]

## Problem B

How many observations (rows) and variables (columns) are in the above dataset? Check the “shape”.

In [20]:
print("The size of train_csv_data is", train_csv_data.shape)

The size of train_csv_data is (1460, 81)


So there are 1460 rows and 81 columns.

## Problem C

What are the data types (“dtypes”) – Numeric/Categorical – of the variables (columns) in the dataset?

In [21]:
train_csv_data.dtypes

Id                 int64
MSSubClass         int64
MSZoning          object
LotFrontage      float64
LotArea            int64
                  ...   
MoSold             int64
YrSold             int64
SaleType          object
SaleCondition     object
SalePrice          int64
Length: 81, dtype: object

## Problem D

What does the .info() method do? Use the .info() method on the imported dataset to check this out.

In [22]:
train_csv_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 81 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1460 non-null   int64  
 1   MSSubClass     1460 non-null   int64  
 2   MSZoning       1460 non-null   object 
 3   LotFrontage    1201 non-null   float64
 4   LotArea        1460 non-null   int64  
 5   Street         1460 non-null   object 
 6   Alley          91 non-null     object 
 7   LotShape       1460 non-null   object 
 8   LandContour    1460 non-null   object 
 9   Utilities      1460 non-null   object 
 10  LotConfig      1460 non-null   object 
 11  LandSlope      1460 non-null   object 
 12  Neighborhood   1460 non-null   object 
 13  Condition1     1460 non-null   object 
 14  Condition2     1460 non-null   object 
 15  BldgType       1460 non-null   object 
 16  HouseStyle     1460 non-null   object 
 17  OverallQual    1460 non-null   int64  
 18  OverallC

## Problem E

What does the .describe() method do? Use the .describe() method on the imported dataset to check.

In [23]:
train_csv_data.describe()

Id   MSSubClass  LotFrontage        LotArea  OverallQual  \
count  1460.000000  1460.000000  1201.000000    1460.000000  1460.000000   
mean    730.500000    56.897260    70.049958   10516.828082     6.099315   
std     421.610009    42.300571    24.284752    9981.264932     1.382997   
min       1.000000    20.000000    21.000000    1300.000000     1.000000   
25%     365.750000    20.000000    59.000000    7553.500000     5.000000   
50%     730.500000    50.000000    69.000000    9478.500000     6.000000   
75%    1095.250000    70.000000    80.000000   11601.500000     7.000000   
max    1460.000000   190.000000   313.000000  215245.000000    10.000000   

       OverallCond    YearBuilt  YearRemodAdd   MasVnrArea   BsmtFinSF1  ...  \
count  1460.000000  1460.000000   1460.000000  1452.000000  1460.000000  ...   
mean      5.575342  1971.267808   1984.865753   103.685262   443.639726  ...   
std       1.112799    30.202904     20.645407   181.066207   456.098091  ...   
min       1.000000  1872.000000   1950.000000     0.000000     0.000000  ...   
25%       5.000000  1954.000000   1967.000000     0.000000     0.000000  ...   
50%       5.000000  1973.000000   1994.000000     0.000000   383.500000  ...   
75%       6.000000  2000.000000   2004.000000   166.000000   712.250000  ...   
max       9.000000  2010.000000   2010.000000  1600.000000  5644.000000  ...   

        WoodDeckSF  OpenPorchSF  EnclosedPorch    3SsnPorch  ScreenPorch  \
count  1460.000000  1460.000000    1460.000000  1460.000000  1460.000000   
mean     94.244521    46.660274      21.954110     3.409589    15.060959   
std     125.338794    66.256028      61.119149    29.317331    55.757415   
min       0.000000     0.000000       0.000000     0.000000     0.000000   
25%       0.000000     0.000000       0.000000     0.000000     0.000000   
50%       0.000000    25.000000       0.000000     0.000000     0.000000   
75%     168.000000    68.000000       0.000000     0.000000     0.000000   
max     857.000000   547.000000     552.000000   508.000000   480.000000   

          PoolArea       MiscVal       MoSold       YrSold      SalePrice  
count  1460.000000   1460.000000  1460.000000  1460.000000    1460.000000  
mean      2.758904     43.489041     6.321918  2007.815753  180921.195890  
std      40.177307    496.123024     2.703626     1.328095   79442.502883  
min       0.000000      0.000000     1.000000  2006.000000   34900.000000  
25%       0.000000      0.000000     5.000000  2007.000000  129975.000000  
50%       0.000000      0.000000     6.000000  2008.000000  163000.000000  
75%       0.000000      0.000000     8.000000  2009.000000  214000.000000  
max     738.000000  15500.000000    12.000000  2010.000000  755000.000000  

[8 rows x 38 columns]

# Problem Set 2

Check Summer Olympic 2016 medal tally : [https://en.wikipedia.org/wiki/2016_Summer_Olympics_medal_table](https://en.wikipedia.org/wiki/2016_Summer_Olympics_medal_table)

## Problem A

Import the Wikipedia page in Jupyter Notebook (check M1 DataAcquisition.ipynb for hints about this).

In [24]:
summer_olympics_website = pd.read_html(
    "https://en.wikipedia.org/wiki/2016_Summer_Olympics_medal_table"
)
print("Type of summer_olympics_website is", type(summer_olympics_website))

Type of summer_olympics_website is <class 'list'>


## Problem B

How many tables are in this Wikipedia page? Check the “len” of the imported data/page to find this out.

In [25]:
print("There are", len(summer_olympics_website), "tables in the Wikipedia page.")

There are 7 tables in the Wikipedia page.


## Problem C

Which one is the actual “2016 Summer Olympics medal table”? Explore all tables in the data to know.

In [26]:
for index, tables in enumerate(summer_olympics_website):
    print("The " + str(index) + "-th table.")
    print(tables.head(3))
    print()

The 0-th table.
  2016 Summer Olympics medals 2016 Summer Olympics medals.1
0                    Location        Rio de Janeiro, Brazil
1                  Highlights                    Highlights
2            Most gold medals            United States (46)

The 1-th table.
                                                   0
0                                Part of a series on
1                               2016 Summer Olympics
2  Bid process (bid details) Development (venues,...

The 2-th table.
  Rank            NOC  Gold  Silver  Bronze  Total
0    1  United States    46      37      38    121
1    2  Great Britain    27      23      17     67
2    3          China    26      18      26     70

The 3-th table.
                                         Ruling date  \
0  List of official changes in medal standings (d...   
1                                     18 August 2016   
2                                     18 August 2016   

                                         Sport/Event

After checking all tables, we could found that the 2-th table is the answer.

In [27]:
summer_olympics_website[2].head()

Rank            NOC  Gold  Silver  Bronze  Total
0    1  United States    46      37      38    121
1    2  Great Britain    27      23      17     67
2    3          China    26      18      26     70
3    4         Russia    19      17      20     56
4    5        Germany    17      10      15     42

## Problem D

Extract the main table, “2016 Summer Olympics medal table”, and store it as a new Pandas DataFrame.

In [28]:
summer_olympics_table = summer_olympics_website[2]
summer_olympics_table.head()

Rank            NOC  Gold  Silver  Bronze  Total
0    1  United States    46      37      38    121
1    2  Great Britain    27      23      17     67
2    3          China    26      18      26     70
3    4         Russia    19      17      20     56
4    5        Germany    17      10      15     42

## Problem E

Extract the TOP 20 countries from the medal table, as above, and store these rows as a new DataFrame.

In [29]:
top_summer_olympics_medal_countries = summer_olympics_table.head(20)
top_summer_olympics_medal_countries

Rank            NOC  Gold  Silver  Bronze  Total
0     1  United States    46      37      38    121
1     2  Great Britain    27      23      17     67
2     3          China    26      18      26     70
3     4         Russia    19      17      20     56
4     5        Germany    17      10      15     42
5     6          Japan    12       8      21     41
6     7         France    10      18      14     42
7     8    South Korea     9       3       9     21
8     9          Italy     8      12       8     28
9    10      Australia     8      11      10     29
10   11    Netherlands     8       7       4     19
11   12        Hungary     8       3       4     15
12   13        Brazil*     7       6       6     19
13   14          Spain     7       4       6     17
14   15          Kenya     6       6       1     13
15   16        Jamaica     6       3       2     11
16   17        Croatia     5       3       2     10
17   18           Cuba     5       2       4     11
18   19    New Zealand     4       9       5     18
19   20         Canada     4       3      15     22

# Bonus Problems

## Problem A

Download the “Census Income” dataset (source : [https://archive.ics.uci.edu/ml/datasets/Census+Income](https://archive.ics.uci.edu/ml/datasets/Census+Income)) from
the UCI Machine Learning Repository (in the “Data Folder”), and import it in Jupyter Notebook as a DataFrame.

Explore the dataset using .shape, .info() and .describe(), exactly as you did in Problem 1 above. Do you spot anything interesting while exploring this dataset? Discuss amongst friends or talk to the Instructor, if you did.

In [30]:
census_income = pd.read_table("adult.data", sep=",", header=None)
census_income.head()

0                  1       2           3   4                    5   \
0  39          State-gov   77516   Bachelors  13        Never-married   
1  50   Self-emp-not-inc   83311   Bachelors  13   Married-civ-spouse   
2  38            Private  215646     HS-grad   9             Divorced   
3  53            Private  234721        11th   7   Married-civ-spouse   
4  28            Private  338409   Bachelors  13   Married-civ-spouse   

                   6               7       8        9     10  11  12  \
0        Adm-clerical   Not-in-family   White     Male  2174   0  40   
1     Exec-managerial         Husband   White     Male     0   0  13   
2   Handlers-cleaners   Not-in-family   White     Male     0   0  40   
3   Handlers-cleaners         Husband   Black     Male     0   0  40   
4      Prof-specialty            Wife   Black   Female     0   0  40   

               13      14  
0   United-States   <=50K  
1   United-States   <=50K  
2   United-States   <=50K  
3   United-States   <=50K  
4            Cuba   <=50K

In [31]:
print("The shape of the data of census_income is", census_income.shape)

The shape of the data of census_income is (32561, 15)


In [32]:
census_income.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   0       32561 non-null  int64 
 1   1       32561 non-null  object
 2   2       32561 non-null  int64 
 3   3       32561 non-null  object
 4   4       32561 non-null  int64 
 5   5       32561 non-null  object
 6   6       32561 non-null  object
 7   7       32561 non-null  object
 8   8       32561 non-null  object
 9   9       32561 non-null  object
 10  10      32561 non-null  int64 
 11  11      32561 non-null  int64 
 12  12      32561 non-null  int64 
 13  13      32561 non-null  object
 14  14      32561 non-null  object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB


In [33]:
census_income.describe()

0             2             4             10            11  \
count  32561.000000  3.256100e+04  32561.000000  32561.000000  32561.000000   
mean      38.581647  1.897784e+05     10.080679   1077.648844     87.303830   
std       13.640433  1.055500e+05      2.572720   7385.292085    402.960219   
min       17.000000  1.228500e+04      1.000000      0.000000      0.000000   
25%       28.000000  1.178270e+05      9.000000      0.000000      0.000000   
50%       37.000000  1.783560e+05     10.000000      0.000000      0.000000   
75%       48.000000  2.370510e+05     12.000000      0.000000      0.000000   
max       90.000000  1.484705e+06     16.000000  99999.000000   4356.000000   

                 12  
count  32561.000000  
mean      40.437456  
std       12.347429  
min        1.000000  
25%       40.000000  
50%       40.000000  
75%       45.000000  
max       99.000000

## Problem B

Note that the Summer Olympic medal tally on Wikipedia follows a really nice structure for the URL, where you can simply change the year in [https://en.wikipedia.org/wiki/2016_Summer_Olympics_medal_table](https://en.wikipedia.org/wiki/2016_Summer_Olympics_medal_table) to fetch any Summer Olympic page. Try changing 2016 in the URL to 2012 or 2008 or 2004 to see for yourself. This allows us to fetch the Olympics medal table from all these years (in fact, any year) quite easily. Let’s try the following.

Write a loop to extract the main tables, “20XX Summer Olympics medal table”, from 2000 to 2016, that is, for the five consecutive Olympics in 2000, 2004, 2008, 2012 and 2016. Store all five tables in respective DataFrames. Now, extract the TOP 20 countries from each of these medal tables, and store these rows as new DataFrames.

In [34]:
summer_olympics_tables_from_2000_to_2016 = {}
for year in range(2000, 2020, 4):
    url = "https://en.wikipedia.org/wiki/" + str(year) + "_Summer_Olympics_medal_table"
    summer_olympics_tables_from_2000_to_2016[year] = pd.read_html(url)[2].head(20)
summer_olympics_tables_from_2000_to_2016[2012]

Rank             NOC  Gold  Silver  Bronze  Total
0     1   United States    47      27      30    104
1     2           China    39      31      22     92
2     3  Great Britain*    29      18      18     65
3     4          Russia    18      21      27     66
4     5     South Korea    13       9       8     30
5     6         Germany    11      20      13     44
6     7          France    11      11      13     35
7     8       Australia     8      15      12     35
8     9           Italy     8       9      11     28
9    10         Hungary     8       4       6     18
10   11           Japan     7      14      17     38
11   12            Iran     7       5       1     13
12   13     Netherlands     6       6       8     20
13   14     New Zealand     6       2       5     13
14   15         Ukraine     5       4      10     19
15   16            Cuba     5       3       7     15
16   17           Spain     4      10       6     20
17   18         Jamaica     4       5       3     12
18   19  Czech Republic     4       3       4     11
19   20    South Africa     4       1       1      6